Mathlib/Algebra/Order/Quantale.lean

```diff
@@ -0,0 +1,160 @@
+* `x ⇨ₗ y` : `sSup { z | z * x ≤ y }`, the leftResiduation of `y` over `x`;
+
+* `x ⇨ᵣ y` : `sSup { z | x * z ≤ y }`, the rightResiduation of `y` over `x`;
+
+## References
+
+<https://en.wikipedia.org/wiki/Quantale>
+<https://encyclopediaofmath.org/wiki/Quantale>
+<https://ncatlab.org/nlab/show/quantale>
+
+-/
+
+/-- An additive quantale is an additive semigroup distributing over a complete lattice. -/
+class AddQuantale (α : Type _) [AddSemigroup α] extends CompleteLattice α where
+  /-- Addition is distributive over join in a quantale -/
+  protected add_sSup_eq_iSup_add (x : α) (s : Set α) : x + sSup s = ⨆ y ∈ s, x + y
+  /-- Addition is distributive over join in a quantale -/
+  protected sSup_add_eq_iSup_add (s : Set α) (y : α) : sSup s + y = ⨆ x ∈ s, x + y
+
+/-- A quantale is a semigroup distributing over a complete lattice. -/
+@[to_additive]
+class Quantale (α : Type _) [Semigroup α] extends CompleteLattice α where
+  /-- Multiplication is distributive over join in a quantale -/
+  protected mul_sSup_eq_iSup_mul (x : α) (s : Set α) : x * sSup s = ⨆ y ∈ s, x * y
+  /-- Multiplication is distributive over join in a quantale -/
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+/-- An integral (or strictly two-sided) additive quantale is a quantale over an additive monoid
+    where top and unit coincide. -/
+class IsIntegralAddQuantale (α : Type _) [AddMonoid α] [AddQuantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_zero : (⊤ : α) = 0
+
+/-- An integral (or strictly two-sided) quantale is a quantale over a monoid where
+    top and unit coincide. -/
+@[to_additive]
+class IsIntegralQuantale (α : Type _) [Monoid α] [Quantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_one : (⊤ : α) = 1
+
+section Quantale
+
+variable (α : Type _)
+variable [Semigroup α] [Quantale α]
+
+@[to_additive]
+theorem mul_sSup_eq_iSup_mul : ∀ x : α, ∀ s : Set α, x * sSup s = ⨆ y ∈ s, x * y :=
+  Quantale.mul_sSup_eq_iSup_mul
+
+@[to_additive]
+theorem sSup_mul_eq_iSup_mul : ∀ s : Set α, ∀ y : α, sSup s * y = ⨆ x ∈ s, x * y :=
+  Quantale.sSup_mul_eq_iSup_mul
+
+end Quantale
+
+section IsIntegral
+open Quantale
+
+variable (α : Type _)
+variable [Monoid α] [Quantale α] [IsIntegralQuantale α]
+
+@[to_additive]
+theorem top_eq_one : (⊤ : α) = 1 := IsIntegralQuantale.top_eq_one
+
+end IsIntegral
+
+namespace Quantale
+
+variable {α : Type _}
+variable [Semigroup α] [Quantale α]
+
+/-- Left- and right- residuation operators on an additive quantale are similar to the Heyting
+operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
+-/
+@[to_additive "Left- and right- residuation operators on an additive quantale are similar to
+the Heyting operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ₗ y = sSup { z | z + x ≤ y }`.
```

Comment:
```suggestion
I.e. `x ⇨ₗ y ≤ z ↔ z + x ≤ y`.
```
and can you add that as a lemma?

Reply:
Adding it as a lemma is probably a good idea. But I think of the sSup version is the actual definition, and so I would put that in the comment as well. Do you have a different opinion there?

Reply:
I stand corrected... I've been reading into residuals a bit more and am convinced you are actually right. Will update.

Reply:
Also, I think it is actually formulated as $z \leq x \Rightarrow\_l y \iff z * x \leq y$.  See also https://www.sciencedirect.com/science/article/pii/S0168007213001553